File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/daylight/DaylightLocationStore.kt
/*
 * DaylightLocationStore.kt
 *
 * Tiny Compose-friendly store the Home `DaylightHero` card reads
 * its lat/long from. Sits in front of `LocationService.captureCurrent`
 * so the hero doesn't have to know about Android location APIs,
 * permission states, or the scan-flow capture path.
 *
 * Mirror of iOS `DaylightLocationStore.swift`.
 *
 * Strategy:
 *   - Persist the last successful fix in SharedPreferences so every
 *     launch after the first paints the hero instantly. Sunrise/
 *     sunset drift by <1 minute over ~100km — persisted coords stay
 *     good for weeks even if the user travels.
 *
 *   - On `refreshIfNeeded`, kick off one suspend `captureCurrent`
 *     and overwrite the cache. Skipped when permission isn't
 *     granted (the hero is not a prompt path; it must not surface
 *     the system dialog).
 *
 *   - In-flight de-dup: a single `isFetching` flag prevents two
 *     overlapping calls when refresh is invoked from both a
 *     `LaunchedEffect` and a lifecycle hook.
 */

package app.quickink.mobile.features.daylight

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.quickink.mobile.features.scan.LocationService
import androidx.core.content.edit

/**
 * Holds the cached lat/long the daylight bar reads from. Hosted
 * by `QuickInkRoot.MainShell` via `remember(applicationContext)`
 * so it persists across recomposition but is bound to the
 * activity's lifetime.
 */
class DaylightLocationStore(private val appContext: Context) {

    /** Latitude of the last successful fix, or null when none yet. */
    var latitude: Double? by mutableStateOf(null)
        private set

    /** Longitude of the last successful fix, or null when none yet. */
    var longitude: Double? by mutableStateOf(null)
        private set

    private val prefs: SharedPreferences =
        appContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private var isFetching: Boolean = false

    init {
        // Seed from the persisted cache — re-launches paint the bar
        // immediately rather than flashing the loading shell.
        if (prefs.contains(KEY_LAT) && prefs.contains(KEY_LNG)) {
            // Stored as Float — sunrise/sunset don't need lat/long
            // beyond ~4 decimal places, and Float gives us that.
            latitude  = prefs.getFloat(KEY_LAT, 0f).toDouble()
            longitude = prefs.getFloat(KEY_LNG, 0f).toDouble()
        }
    }

    /**
     * Kick off a one-shot location fetch if we don't already have
     * coordinates AND permission is granted. No-op when there's
     * already an in-flight fetch, no permission, or the cache is
     * still fresh (<24h). Safe to call repeatedly from a
     * `LaunchedEffect`.
     */
    suspend fun refreshIfNeeded() {
        if (isFetching) return
        if (!LocationService.hasPermission(appContext)) return
        if (latitude != null && longitude != null && !cacheIsStale()) return

        isFetching = true
        try {
            val captured = LocationService.captureCurrent(appContext) ?: return
            latitude  = captured.latitude
            longitude = captured.longitude
            prefs.edit {
                putFloat(KEY_LAT, captured.latitude.toFloat())
                putFloat(KEY_LNG, captured.longitude.toFloat())
                putLong (KEY_FETCHED_AT, System.currentTimeMillis())
            }
        } finally {
            isFetching = false
        }
    }

    /**
     * Cache considered stale after 24h. The sun's path is the
     * same week to week at a fixed point on Earth; a daily refresh
     * keeps us honest if the user travelled overnight.
     */
    private fun cacheIsStale(): Boolean {
        val ts = prefs.getLong(KEY_FETCHED_AT, 0L)
        if (ts == 0L) return true
        val ageMs = System.currentTimeMillis() - ts
        return ageMs > 24L * 60L * 60L * 1000L
    }

    private companion object {
        const val PREFS_NAME     = "quickink.daylight"
        const val KEY_LAT        = "lat"
        const val KEY_LNG        = "lng"
        const val KEY_FETCHED_AT = "fetched_at"
    }
}

File: apps/quickink/ios/QuickInk/Calendar/DaylightLocationStore.swift
/*
 * DaylightLocationStore.swift
 *
 * Tiny @ObservableObject the Home `DaylightHero` card reads its
 * lat/long from. Sits in front of `LocationService.captureCurrent()`
 * so the hero doesn't have to know about CoreLocation, permission
 * states, or the scan-flow capture path.
 *
 * Strategy:
 *   - Persist the last successful fix in `UserDefaults` so every
 *     launch after the first paints the hero instantly (sunrise/
 *     sunset drift by <1 minute over ~100km — the persisted coords
 *     stay good for weeks even if the user travels). Solar math is
 *     latitude-dominant; longitude only matters for the time-zone
 *     shift, which is independent of this cache.
 *
 *   - On `refreshIfNeeded`, kick off one async `captureCurrent`
 *     and overwrite the cache. Skipped when permission isn't
 *     granted (this isn't a prompt path; the hero shouldn't trigger
 *     a system dialog).
 *
 *   - In-flight de-dup: a single `fetchTask` slot prevents two
 *     overlapping calls when `refreshIfNeeded` is invoked from
 *     both `QuickInkRoot.task` and a scene-active hook.
 */

import Foundation

@MainActor
public final class DaylightLocationStore: ObservableObject {

    @Published public private(set) var latitude:  Double?
    @Published public private(set) var longitude: Double?

    private var fetchTask: Task<Void, Never>?

    private enum Keys {
        static let lat = "quickink.daylight.lat"
        static let lng = "quickink.daylight.lng"
    }

    public init() {
        // Seed from the persisted cache — re-launches paint the bar
        // immediately rather than flashing the loading shell.
        let defaults = UserDefaults.standard
        if defaults.object(forKey: Keys.lat) != nil,
           defaults.object(forKey: Keys.lng) != nil {
            latitude  = defaults.double(forKey: Keys.lat)
            longitude = defaults.double(forKey: Keys.lng)
        }
    }

    /// Kick off a one-shot location fetch if we don't already have
    /// coordinates AND permission is granted. No-op when there's
    /// already an in-flight task or the user hasn't granted location.
    /// Safe to call repeatedly (e.g. on every `scenePhase == .active`).
    public func refreshIfNeeded() {
        guard fetchTask == nil else { return }
        let status = LocationService.shared.authorizationStatus
        guard status == .authorizedWhenInUse || status == .authorizedAlways else {
            return
        }
        // Skip the refetch when we already have a cache AND a fetch
        // happened recently — keeps the bar from waking GPS on every
        // foreground.
        if latitude != nil, longitude != nil,
           !Self.cacheIsStale() { return }

        fetchTask = Task { [weak self] in
            defer { self?.fetchTask = nil }
            guard let captured = await LocationService.shared.captureCurrent() else {
                return
            }
            self?.applyAndPersist(latitude: captured.latitude, longitude: captured.longitude)
        }
    }

    private func applyAndPersist(latitude: Double, longitude: Double) {
        self.latitude  = latitude
        self.longitude = longitude
        let defaults = UserDefaults.standard
        defaults.set(latitude,  forKey: Keys.lat)
        defaults.set(longitude, forKey: Keys.lng)
        defaults.set(Date().timeIntervalSince1970, forKey: "quickink.daylight.fetched_at")
    }

    /// Cache considered stale after 24h. The sun's path is the same
    /// week to week at a fixed point on Earth; a daily refresh keeps
    /// us honest if the user travelled overnight.
    private static func cacheIsStale() -> Bool {
        let key = "quickink.daylight.fetched_at"
        let ts  = UserDefaults.standard.double(forKey: key)
        guard ts > 0 else { return true }
        let age = Date().timeIntervalSince1970 - ts
        return age > 24 * 60 * 60
    }
}
